components/external-api-framework/org.wso2.carbon.identity.external.api.client/src/main/java/org/wso2/carbon/identity/external/api/client/internal/service/APIClient.java

```diff
@@ -0,0 +1,202 @@
+        switch (requestContext.getHttpMethod()) {
+            case POST:
+                httpEntityEnclosingRequestBase = new HttpPost(requestContext.getEndpointUrl());
+                break;
+            default:
+                throw new APIClientInvocationException(
+                        ErrorMessage.ERROR_CODE_UNSUPPORTED_HTTP_METHOD, requestContext.getHttpMethod().getName());
+        }
+        setRequestEntity(httpEntityEnclosingRequestBase, requestContext);
+
+        try {
+            return executeRequest(httpEntityEnclosingRequestBase, apiInvocationConfig);
+        } finally {
+            httpEntityEnclosingRequestBase.releaseConnection();
+        }
+    }
+
+    private void setRequestEntity(HttpEntityEnclosingRequestBase httpRequestBase, APIRequestContext requestContext) {
+
+        StringEntity entity = new StringEntity(requestContext.getPayload(), StandardCharsets.UTF_8);
+        httpRequestBase.setEntity(entity);
+
+        // only be default
+        httpRequestBase.setHeader("Accept", "application/json");
+        httpRequestBase.setHeader("Content-type", "application/json");
+        Header authHeader = APIRequestBuildingUtils.buildAuthenticationHeader(requestContext.getApiAuthentication());
+        if (authHeader != null) {
+            httpRequestBase.setHeader(authHeader);
+        }
+        for (Map.Entry<String, String> header : requestContext.getHeaders().entrySet()) {
+            httpRequestBase.setHeader(header.getKey(), header.getValue());
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Request entity configured successfully for endpoint: %s with payload length: " +
+                            "%d and headers count: %d",
+                    requestContext.getEndpointUrl(),
+                    requestContext.getPayload() != null ? requestContext.getPayload().length() : 0,
+                    requestContext.getHeaders().size()
+            ));
+        }
+    }
+
+    private APIResponse executeRequest(HttpEntityEnclosingRequestBase request, APIInvocationConfig apiInvocationConfig)
+            throws APIClientInvocationException {
+
+
+        int allowedRetryCount = apiInvocationConfig.getAllowedRetryCount();
+        int attempt = 0;
+        while (true) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Executing request to URI: %s, retry attempt: %d/%d",
+                        request.getURI(), attempt, apiInvocationConfig.getAllowedRetryCount()
+                ));
+            }
+
+            try (CloseableHttpResponse response = httpClient.execute(request)) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug(String.format("Request executed successfully to URI: %s, response status: %d",
+                            request.getURI(), response.getStatusLine().getStatusCode()));
+                }
+                return handleResponse(response);
+            } catch (IOException e) {
```

Comment:
The error exit only happen for IO exception under an infinite loop. Shall we make this generic Exception ? WDYT ? Or else you may use another exit point for anything else.

Reply:
If there is different type exception, then method exits immediately. right?